### PR TITLE
chore(har): allow replaying from zip har

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: packages/playwright-core/src/client/harRouter.ts [hunk 1]
```diff
@@ -46,15 +53,33 @@ export class HarRouter {
...
         debugLogger.log('api', e);
       }
+
       if (response) {
         debugLogger.log('api', `serving from HAR: ${route.request().method()} ${route.request().url()}`);
-        await route.fulfill({ response });
-      } else if (options?.fallback === 'continue') {
+        const sha1 = (response.content as any)._sha1;
+        if (zipFile && sha1) {
+          const body = await zipFile.read(sha1).catch(e => null);
```
Comment (yury-s) [inline packages/playwright-core/src/client/harRouter.ts]: log the error

File: packages/playwright-core/src/client/harRouter.ts [hunk 2]
```diff
@@ -18,48 +18,85 @@ import fs from 'fs';
...
 export class HarRouter {
   private _pattern: string | RegExp;
-  private _handler: (route: Route) => Promise<void>;
+  private _harFile: HARFile;
+  private _zipFile: ZipFile | null;
+  private _options: HarOptions | undefined;
 
   static async create(options: HarOptions): Promise<HarRouter> {
+    if (options.path.endsWith('.zip')) {
+      const zipFile = new ZipFile(options.path);
```
Comment (mxschmitt) [inline packages/playwright-core/src/client/harRouter.ts:37]: localUtils instead?

Comment (pavelfeldman) [reply packages/playwright-core/src/client/harRouter.ts:37]: In the follow up, yes

**Review by dgozman — APPROVED** [review]

File: packages/playwright-core/src/client/harRouter.ts [hunk 3]
```diff
@@ -29,11 +30,17 @@ export class HarRouter {
   private _handler: (route: Route) => Promise<void>;
 
   static async create(options: HarOptions): Promise<HarRouter> {
+    if (options.path.endsWith('.zip')) {
+      const zipFile = new ZipFile(options.path);
+      const har = await zipFile.read('har.har');
```
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply packages/playwright-core/src/client/harRouter.ts:38]: Done.

**Pull request merged**